include/seqan3/contrib/parallel/latch.hpp

```diff
@@ -0,0 +1,164 @@
+        assert(expected >= 0);
+        num_waiting.store(0, std::memory_order_relaxed);
+    }
+    //!\}
+
+    /*!\brief Atomically decrements counter by `n`.
+     * \param n The value to subtract.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point with `n` count without waiting. If `n` is less than 0 or greater then the
+     * current count, then this operation results in undefined behaviour for all other participating threads.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void arrive(ptrdiff_t n = 1) noexcept
+    {
+        assert(counter.load(std::memory_order_acquire) >= n);
+        assert(counter.load(std::memory_order_acquire) >= 0);
+
+        counter.fetch_sub(n, std::memory_order_acq_rel);
+    }
+
+    /*!\brief Atomically decrements counter by `n` and blocks the calling thread.
+     * \param n The value to subtract.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point with `n` count and blocks the calling thread until all participating threads
+     * have arrived. If `n` is less than 0 or greater then the current count, then this operation results in undefined
+     * behaviour for all other participating threads.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void arrive_and_wait(ptrdiff_t n = 1) noexcept
+    {
+        ++num_waiting; // ensure that destructor is not finished in-between the arrive and wait call.
+        arrive(n);
+        wait();
+        --num_waiting;
+    }
+
+    /*!\brief Checks if all participating threads have reached the synchronisation point.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    bool try_wait() const noexcept
+    {
+        return counter.load(std::memory_order_acquire) == 0;
+    }
+
+    /*!\brief Waits for all participating threads to arrive at the synchronisation point.
+     *
+     * \details
+     *
+     * If counter == 0 returns immediately, otherwise blocks the calling thread at the synchronisation point.
+     * This implementation uses a seqan3::spin_delay to wait for the value.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void wait() const
+    {
+        ++num_waiting; // register waiting thread to synchronise with destructor.
+        spin_delay delay{};
+        while (counter.load(std::memory_order_acquire) > 0)
+            delay.wait();
+        --num_waiting;
+    }
+
+private:
+
+    //!\brief The number of participating threads.
+    alignas(std::hardware_destructive_interference_size) std::atomic<std::ptrdiff_t>         counter;
+    //!\brief The number of waiting threads.
+    alignas(std::hardware_destructive_interference_size) mutable std::atomic<std::ptrdiff_t> num_waiting;
```

Comment:
yeah, that's how you do const correctness :yellow_heart:

Reply:
I am merely following the proposal interface 😄